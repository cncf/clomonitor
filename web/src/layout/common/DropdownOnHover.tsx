import classNames from 'classnames';
import { isUndefined } from 'lodash';
import { useEffect, useRef, useState } from 'react';

import useOutsideClick from '../../hooks/useOutsideClick';
import styles from './DropdownOnHover.module.css';

interface Props {
  linkContent: JSX.Element | string;
  children: JSX.Element;
  className?: string;
}

const DropdownOnHover = (props: Props) => {
  const ref = useRef(null);
  const [openStatus, setOpenStatus] = useState(false);
  const [onLinkHover, setOnLinkHover] = useState(false);
  const [onDropdownHover, setOnDropdownHover] = useState(false);
  useOutsideClick([ref], openStatus, () => setOpenStatus(false));

  useEffect(() => {
    let timeout: NodeJS.Timeout;
    if (!openStatus && (onLinkHover || onDropdownHover)) {
      timeout = setTimeout(() => {
        setOpenStatus(true);
      }, 100);
    }
    if (openStatus && !onLinkHover && !onDropdownHover) {
      timeout = setTimeout(() => {
        // Delay to hide the dropdown to let some time for changing between dropdown and link
        setOpenStatus(false);
      }, 50);
    }
    return () => {
      if (!isUndefined(timeout)) {
        clearTimeout(timeout);
      }
    };
  }, [onLinkHover, onDropdownHover, openStatus]);

  return (
    <>
      <div className={props.className}>
        <div className="position-absolute">
          <div
            ref={ref}
            role="complementary"
            className={classNames('dropdown-menu rounded-0 text-wrap', styles.dropdown, {
              show: openStatus,
            })}
            onMouseEnter={() => setOnDropdownHover(true)}
            onMouseLeave={() => setOnDropdownHover(false)}
          >
            <div className="px-3 py-1">{props.children}</div>
          </div>
        </div>

        <div
          onMouseEnter={(e) => {
            e.preventDefault();
            setOnLinkHover(true);
          }}
          onMouseLeave={() => {
            setOnLinkHover(false);
          }}
          aria-expanded={openStatus}
        >
          {props.linkContent}
        </div>
      </div>
    </>
  );
};

export default DropdownOnHover;
