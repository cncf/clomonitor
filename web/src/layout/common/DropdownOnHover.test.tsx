import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import DropdownOnHover from './DropdownOnHover';

describe('ParamInfo', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <DropdownOnHover linkContent="content">
        <>children</>
      </DropdownOnHover>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    it('renders component', () => {
      render(
        <DropdownOnHover linkContent="content">
          <>children</>
        </DropdownOnHover>
      );

      expect(screen.getByText('content')).toBeInTheDocument();
      expect(screen.getByRole('complementary')).toBeInTheDocument();
      expect(screen.getByRole('complementary')).toHaveTextContent('children');
    });

    it('displays dropdown to enter on content and hides on leave', async () => {
      jest.useFakeTimers();

      render(
        <DropdownOnHover linkContent="content">
          <>children</>
        </DropdownOnHover>
      );

      const dropdown = screen.getByRole('complementary');

      expect(dropdown).not.toHaveClass('show');

      userEvent.hover(screen.getByText('content'));

      act(() => {
        jest.advanceTimersByTime(100);
      });

      expect(dropdown).toHaveClass('show');

      userEvent.unhover(screen.getByText('content'));

      act(() => {
        jest.advanceTimersByTime(50);
      });

      expect(dropdown).not.toHaveClass('show');

      jest.useRealTimers();
    });

    it('hides dropdown to leave it', async () => {
      jest.useFakeTimers();

      render(
        <DropdownOnHover linkContent="content">
          <>children</>
        </DropdownOnHover>
      );

      const dropdown = screen.getByRole('complementary');

      userEvent.hover(screen.getByText('content'));
      userEvent.hover(dropdown);
      userEvent.unhover(screen.getByText('content'));

      act(() => {
        jest.advanceTimersByTime(100);
      });

      expect(dropdown).toHaveClass('show');

      userEvent.unhover(dropdown);

      act(() => {
        jest.advanceTimersByTime(50);
      });

      expect(dropdown).not.toHaveClass('show');

      jest.useRealTimers();
    });
  });
});
