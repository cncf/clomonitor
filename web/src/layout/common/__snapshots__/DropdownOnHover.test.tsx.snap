// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParamInfo creates snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="position-absolute"
    >
      <div
        class="dropdown-menu rounded-0 text-wrap dropdown"
        role="complementary"
        style="width: auto;"
      >
        <div
          class="px-3 py-1"
        >
          children
        </div>
      </div>
    </div>
    <div
      aria-expanded="false"
      class="cursorDefault"
    >
      content
    </div>
  </div>
</DocumentFragment>
`;
